pkg/objects/github.go

```diff
@@ -0,0 +1,277 @@
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/go-github/v44/github"
+	"github.com/majewsky/schwift"
+	"github.com/sapcc/go-bits/secrets"
+	"golang.org/x/oauth2"
+
+	"github.com/sapcc/swift-http-import/pkg/util"
+)
+
+type GithubReleaseSource struct {
+	// Options from config file.
+	URLString         string               `yaml:"url"`
+	Token             secrets.AuthPassword `yaml:"token"`
+	TagNamePattern    string               `yaml:"tag_name_pattern"`
+	IncludeDraft      bool                 `yaml:"include_draft"`
+	IncludePrerelease bool                 `yaml:"include_prerelease"`
+
+	// Compiled configuration.
+	url    *url.URL       `yaml:"-"`
+	client *github.Client `yaml:"-"`
+	// httpClient is an unauthenticated default http.Client that is used for downloading release assets.
+	// We use a separate client instead of the same http.Client that we create and pass to
+	// the github.Client because that http.Client, when obtained using oauth2.NewClient(),
+	// does not return all headers in the request response.
+	httpClient *http.Client   `yaml:"-"`
+	owner      string         `yaml:"-"` // repository owner
+	repo       string         `yaml:"-"`
+	tagNameRx  *regexp.Regexp `yaml:"-"`
+	// notOlderThan is used to limit release listing to prevent excess API requests.
+	notOlderThan *time.Time `yaml:"-"`
+}
+
+// githubRepoRx is used to extract repository owner and name from a URL.
+//
+// Example:
+//   Input: https://github.com/sapcc/swift-http-import
+//   Match groups: ["sapcc", "swift-http-import"]
+var githubRepoRx = regexp.MustCompile(`^https?://[^\s/]+/([^\s/]+)/([^\s/]+)/$`)
+
+// Validate implements the Source interface.
+func (s *GithubReleaseSource) Validate(name string) []error {
+	mL := githubRepoRx.FindStringSubmatch(s.URLString)
+	if mL == nil {
+		return []error{fmt.Errorf("invalid value for %s.url: expected a url in the format %q, got: %q",
+			name, "http(s)://<hostname>/<owner>/<repo>/", s.URLString)}
+	}
+	s.owner = mL[1]
+	s.repo = mL[2]
+
+	url, err := parseURL(s.URLString)
+	if err != nil {
+		return []error{fmt.Errorf("could not parse %s.url: %s", name, err.Error())}
+	}
+	s.url = url
+	if s.url.Hostname() != "github.com" {
+		if s.Token == "" {
+			return []error{fmt.Errorf("%s.token is required for repositories hosted on GitHub Enterprise", name)}
+		}
+	}
+
+	if s.TagNamePattern != "" {
+		s.tagNameRx, err = regexp.Compile(s.TagNamePattern)
+		if err != nil {
+			return []error{fmt.Errorf("could not parse %s.tag_name_pattern: %s", name, err.Error())}
+		}
+	}
+
+	return nil
+}
+
+// Connect implements the Source interface.
+func (s *GithubReleaseSource) Connect(name string) error {
+	// HTTP client for downloading release assets.
+	// Note: we don't need an authenticated client here, we'll just pass the token in the
+	// `Authorization` header when we download the asset.
+	s.httpClient = &http.Client{}
+
+	// Use a separate http.Client for creating github.Client.
+	client2 := &http.Client{}
```

Comment:
By the same argument, I'd prefer DefaultClient here if possible:
```suggestion
	client2 := http.DefaultClient
```